### PR TITLE
(140) new book category

### PR DESCRIPTION
**Pull request by pankaj-ag**
#140
Create new category while assinging them to books.
Added new books category section to edit, delete and create the new category system.
Added permissions for book category.
Assign permissions to super admin only

### CODE REVIEW

**Comment by rathorevaibhav** [comment]
@pankaj-ag please add proper status labels to let the team know about the PR's status.

**Comment by rathorevaibhav** [comment]
@pankaj-ag Can't complete the PR review. I've asked @bohraji to take over. Assigning him as the reviewer.

File: resources/views/knowledgecafe/library/menu.blade.php [hunk 1]
```diff
@@ -2,4 +2,8 @@
     <li class="nav-item">
     <a class="nav-item nav-link {{ $active === 'books' ? 'active' : '' }}" href="{{ route('books.index') }}"><i class="fa fa-book"></i>&nbsp;Books</a>
     </li>
+
+    <li class="nav-item">
+        <a class="nav-item nav-link {{ $active === 'book_category' ? 'active' : '' }}" href="{{ route('books.category.index') }}"><i class="fa fa-book"></i>&nbsp;Books Category</a>
```
Comment (bohraji) [inline resources/views/knowledgecafe/library/menu.blade.php]: With named routes, we do not need to have a separate variable to set active state on menu items. We can make use route functions to do so. Check updates to HR module `menu` and update accordingly to maintain consistency.

File: resources/views/knowledgecafe/library/categories/index.blade.php [hunk 2]
```diff
@@ -0,0 +1,83 @@
+@extends('layouts.app')
+
+@section('content')
+<div class="container" id="books_category">
+    <br>
+    @include('knowledgecafe.library.menu', ['active' => 'book_category'])
```
Comment (bohraji) [inline resources/views/knowledgecafe/library/categories/index.blade.php]: No need to send extra variable to set active mode on menu item. Check updates to HR module

File: resources/views/knowledgecafe/library/categories/index.blade.php [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    <br><br>
+
+    <div class="row">
+        <div class="col-md-6"><h1>Book Category</h1></div>
+        <div class="col-md-6">
+            <a href="#" @click="updateNewCategoryMode('add')" class="btn btn-success float-right">Add Category</a>
```
Comment (bohraji) [inline resources/views/knowledgecafe/library/categories/index.blade.php]: Please remove the empty `href`. We can use a `button` here.

File: resources/assets/sass/app.scss [hunk 4]
```diff
@@ -308,4 +308,23 @@ input:focus + .slider {
             width: 100%;
         }
     }
+}
+
+
+#category_container {
+    .row {
```
Comment (bohraji) [inline resources/assets/sass/app.scss]: The row component default margin is a property of bootstrap and is common across the whole platform. Not sure if changing it in a single module will change the appearance of this module from  the rest of the platform. Updates?

File: resources/views/knowledgecafe/library/categories/index.blade.php [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                <span>No book for this category</span> 
+            </div>
+            
+            <div v-if = "category.assign_books_count" class="col-4">
+                <span> @{{ category.assign_books_count }}</span> 
+                <span> @{{ (category.assign_books_count > 1) ? 'books' : 'book' }} </span>
+            </div>
+
+            <div @click="showEditMode(index)" class="col-2">
+                <a href="#">
```
Comment (bohraji) [inline resources/views/knowledgecafe/library/categories/index.blade.php]: Remove empty `href`

File: resources/views/knowledgecafe/library/categories/index.blade.php [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                    <i class="fa fa-pencil"></i>&nbsp;<u>Edit</u>
+                </a>
+            </div>
+
+            <div class="col-2" @click="deleteCategory(index)">
+                <a href="#">
```
Comment (bohraji) [inline resources/views/knowledgecafe/library/categories/index.blade.php]: Remove empty `href`

File: resources/views/knowledgecafe/library/categories/index.blade.php [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+                    @{{ category.name }}
+                </span>
+
+            </div>
+
+            <div v-if = "!category.assign_books_count" class="col-4">
+                <span>No book for this category</span> 
+            </div>
+            
+            <div v-if = "category.assign_books_count" class="col-4">
```
Comment (bohraji) [inline resources/views/knowledgecafe/library/categories/index.blade.php]: This can be the else of the above statement.

File: routes/web.php [hunk 8]
```diff
@@ -62,6 +62,9 @@
     Route::get('/knowledgecafe', 'KnowledgeCafe\KnowledgeCafeController@index');
     Route::resource('/knowledgecafe/library/books', 'KnowledgeCafe\Library\BookController')
                 ->names([ 'index' => 'books.index', 'create' => 'books.create', 'show' => 'books.show', 'store' => 'books.store']);
-
     Route::post('/knowledgecafe/library/book/fetchinfo', 'KnowledgeCafe\Library\BookController@fetchBookInfo')->name('books.fetchInfo');
+
+    Route::resource('/knowledgecafe/library/book-categories', 'KnowledgeCafe\Library\BookCategoryController')
```
Comment (bohraji) [inline routes/web.php]: All the `/knowledgecafe/library` routes can grouped together in a single namespace. Check updates to `hr` module.

**Comment by bohraji** [comment]
@pankaj-ag Also update the base branch from develop to master. Also, see that the next PR are smaller in size.

**Comment by bohraji** [comment]
Checked this PR :

- [ ] For duplication.
- [ ] For performance. 
- [ ] Usage of recursive methods, loops and algorithms.
- [ ] For readability and understanding. 
- [ ] For functions with multiple responsibilities.
- [ ] For scope of refactoring. 
- [ ] For coupling
- [ ] For alternative approaches to the implementation. 
- [ ] For how the change impacts other parts of the system.
- [ ] For validations. client side and server side validations are present.

For more insights on code reviews checkout our [guidelines](https://docs.google.com/document/d/1Sa-XYJEmGwrOM5gk5zoan_Z6j90ylibmCmt331361nw/edit#).

File: resources/views/knowledgecafe/library/menu.blade.php (hunk 1 above)
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/menu.blade.php]: @bohraji I have some doubts about the new implementation. I think we should track this in a different issue.

File: resources/views/knowledgecafe/library/categories/index.blade.php (hunk 2 above)
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/categories/index.blade.php]: Same as above

File: resources/assets/sass/app.scss (hunk 4 above)
Comment (pankaj-ag) [reply resources/assets/sass/app.scss]: Yes, @bohraji I am aware of that I have to do this to maintain the consistency.

File: app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    /**
+     * Display a listing of the resource.
+     *
+     * @return \Illuminate\View\View
+     */
+    public function index()
+    {
+        $categories = BookCategory::with('books')->orderBy('name')->get();
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php]: Done

File: app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php [hunk 10]
```diff
@@ -0,0 +1,111 @@
+<?php
+
+namespace App\Http\Controllers\KnowledgeCafe\Library;
+
+use App\Models\KnowledgeCafe\Library\BookCategory;
+use Illuminate\Http\Request;
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php]: Removed

File: app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Update the specified resource in storage.
+     *
+     * @param  App\Http\Requests\KnowledgeCafe\Library\BookCategoryRequest  $request
+     * @param  \App\Models\KnowledgeCafe\Library\BookCategory  $bookCategory
+     * @return \Illuminate\Http\Response
+     */
+    public function update(BookCategoryRequest $request, BookCategory $bookCategory)
+    {
+        return response()->json(['success' => $bookCategory->update($request->validated()) ]);
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php]: Chnaged

File: app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * Remove the specified resource from storage.
+     *
+     * @param  \App\Models\KnowledgeCafe\Library\BookCategory  $bookCategory
+     * @return void
+     */
+    public function destroy(BookCategory $bookCategory)
+    {
+        return response()->json(['success' => $bookCategory->delete() ]);
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php]: Done

File: app/Http/Controllers/KnowledgeCafe/Library/BookController.php [hunk 13]
```diff
@@ -25,7 +25,7 @@ public function index()
     {
         $this->authorize('list', Book::class);
         $books = Book::with('categories')->orderBy('title')->get();
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookController.php]: done

File: app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php (hunk 9 above)
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php]: I'll do that after search functionality.

File: resources/views/knowledgecafe/library/categories/index.blade.php (hunk 3 above)
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/categories/index.blade.php]: done

File: resources/views/knowledgecafe/library/categories/index.blade.php (hunk 5 above)
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/categories/index.blade.php]: done

File: resources/views/knowledgecafe/library/categories/index.blade.php (hunk 6 above)
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/categories/index.blade.php]: done

File: resources/views/knowledgecafe/library/categories/index.blade.php (hunk 7 above)
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/categories/index.blade.php]: done

File: routes/web.php (hunk 8 above)
Comment (pankaj-ag) [reply routes/web.php]: Done.

File: app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+     * Update the specified resource in storage.
+     *
+     * @param  App\Http\Requests\KnowledgeCafe\Library\BookCategoryRequest  $request
+     * @param  \App\Models\KnowledgeCafe\Library\BookCategory  $bookCategory
+     * @return \Illuminate\Http\Response
+     */
+    public function update(BookCategoryRequest $request, BookCategory $bookCategory)
+    {
+        return response()->json(
+            ['isUpdatesSuccessfully' => $bookCategory->update($request->validated()) ]
```
Comment (rathorevaibhav) [inline app/Http/Controllers/KnowledgeCafe/Library/BookCategoryController.php]: is this supposed to be `isUpdatedSuccessfully`? If this is supposed to check a boolean, I think  `isUpdated` will make more sense.

File: resources/views/knowledgecafe/library/categories/index.blade.php [hunk 15]
```diff
@@ -0,0 +1,81 @@
+@extends('layouts.app')
+
+@section('content')
+<div class="container" id="books_category" v-cloak>
+    <br>
+    @include('knowledgecafe.library.menu', ['active' => 'book_category'])
```
[the remark this answers is not in the document]
Comment (pankaj-ag) [reply resources/views/knowledgecafe/library/categories/index.blade.php]: We are tracking this in a new issue #301

**Pull request merged**